### ISSUE TITLE
Lots of elidable loop code generated on CPU.

### ISSUE THREAD

**Issue by benvanik** [google/iree#8053]
Repro:
```mlir
builtin.func @main(%arg0: tensor<1x320x320x3xi8>) -> tensor<1x320x320x3xi8> {
  %0 = "tosa.rescale"(%arg0) {double_round = false, input_zp = 128 : i32, multiplier = [1073741824 : i32], output_zp = 0 : i32, per_channel = false, scale32 = true, shift = [30 : i32]} : (tensor<1x320x320x3xi8>) -> tensor<1x320x320x3xi8>
  return %0 : tensor<1x320x320x3xi8>
}
```

This produces the following dispatch:
```mlir

  stream.executable private @main_dispatch_0 {
    stream.executable.export public @main_dispatch_0
    builtin.module  {
      func @main_dispatch_0(%arg0: !stream.binding {stream.alignment = 32 : index}, %arg1: !stream.binding {stream.alignment = 32 : index}) {
        %c3 = arith.constant 3 : index
        %c320 = arith.constant 320 : index
        %c128_i32 = arith.constant 128 : i32
        %c1073741824_i32 = arith.constant 1073741824 : i32
        %c30_i8 = arith.constant 30 : i8
        %c-128_i32 = arith.constant -128 : i32
        %c127_i32 = arith.constant 127 : i32
        %c0 = arith.constant 0 : index
        %0 = stream.binding.subspan %arg0[%c0] : !stream.binding -> !flow.dispatch.tensor<readonly:320x320x3xi8>
        %1 = stream.binding.subspan %arg1[%c0] : !stream.binding -> !flow.dispatch.tensor<writeonly:320x320x3xi8>
        %workgroup_size_0 = flow.dispatch.workgroup.size[0] : index
        %workgroup_size_1 = flow.dispatch.workgroup.size[1] : index
        %workgroup_size_2 = flow.dispatch.workgroup.size[2] : index
        %workgroup_id_0 = flow.dispatch.workgroup.id[0] : index
        %workgroup_count_0 = flow.dispatch.workgroup.count[0] : index
        %workgroup_id_1 = flow.dispatch.workgroup.id[1] : index
        %workgroup_count_1 = flow.dispatch.workgroup.count[1] : index
        %workgroup_id_2 = flow.dispatch.workgroup.id[2] : index
        %workgroup_count_2 = flow.dispatch.workgroup.count[2] : index
        %2 = affine.apply #map0()[%workgroup_id_2, %workgroup_size_2]
        %3 = affine.apply #map0()[%workgroup_count_2, %workgroup_size_2]
        scf.for %arg2 = %2 to %c320 step %3 {
          %4 = affine.apply #map0()[%workgroup_id_1, %workgroup_size_1]
          %5 = affine.apply #map0()[%workgroup_count_1, %workgroup_size_1]
          scf.for %arg3 = %4 to %c320 step %5 {
            %6 = affine.apply #map0()[%workgroup_id_0, %workgroup_size_0]
            %7 = affine.apply #map0()[%workgroup_count_0, %workgroup_size_0]
            scf.for %arg4 = %6 to %c3 step %7 {
              %8 = affine.min #map1(%arg2)[%workgroup_size_2]
              %9 = affine.min #map1(%arg3)[%workgroup_size_1]
              %10 = affine.min #map2(%arg4)[%workgroup_size_0]
              %11 = flow.dispatch.tensor.load %0, offsets = [%arg2, %arg3, %arg4], sizes = [%8, %9, %10], strides = [1, 1, 1] : !flow.dispatch.tensor<readonly:320x320x3xi8> -> tensor<?x?x?xi8>
              %12 = linalg.init_tensor [%8, %9, %10] : tensor<?x?x?xi8>
              %13 = linalg.generic {indexing_maps = [#map3, #map3], iterator_types = ["parallel", "parallel", "parallel"]} ins(%11 : tensor<?x?x?xi8>) outs(%12 : tensor<?x?x?xi8>) {
              ^bb0(%arg5: i8, %arg6: i8):  // no predecessors
                %14 = arith.extsi %arg5 : i8 to i32
                %15 = arith.subi %14, %c128_i32 : i32
                %16 = "tosa.apply_scale"(%15, %c1073741824_i32, %c30_i8) {double_round = false} : (i32, i32, i8) -> i32
                %17 = arith.cmpi slt, %16, %c-128_i32 : i32
                %18 = select %17, %c-128_i32, %16 : i32
                %19 = arith.cmpi slt, %c127_i32, %16 : i32
                %20 = select %19, %c127_i32, %18 : i32
                %21 = arith.trunci %20 : i32 to i8
                linalg.yield %21 : i8
              } -> tensor<?x?x?xi8>
              flow.dispatch.tensor.store %13, %1, offsets = [%arg2, %arg3, %arg4], sizes = [%8, %9, %10], strides = [1, 1, 1] : tensor<?x?x?xi8> -> !flow.dispatch.tensor<writeonly:320x320x3xi8>
            }
          }
        }
        return
      }
    }
  }
```

Which lowers into quite a bit of code through LLVM and a pretty complex function:
![image](https://user-images.githubusercontent.com/75337/148658271-4bd0dcb8-d6da-4869-a8d2-eec8655560d5.png)

Unrelated (but interesting), LLVM also produces weird empty spin loops:
```asm
.LBB0_27:
	addq	%r8, %rcx
	cmpq	$320, %rcx
	jl	.LBB0_27
```

**Comment by benvanik** [comment]
If hand-authored, the red blocks would be the only ones present (parameter unpack/setup, core loop, exit), so a reasonable target to shoot for: 
![image](https://user-images.githubusercontent.com/75337/148658451-ba2af032-6a4d-4476-89a7-bca9cdd4189c.png)

**Comment by nicolasvasilache** [comment]
nice, could you elaborate a bit on how you got this viz?

**Comment by benvanik** [comment]
https://cutter.re/ loading one of the shared libraries we spit out (with `-iree-llvm-keep-linker-artifacts`) - since the .so's only contain the codegened functions they are easy to pick out and inspect via the graph tab; the decompiler is also a good way to see the control flow (I find it much more useful than looking at LLVM IR dumps and such): 
![image](https://user-images.githubusercontent.com/75337/153269681-06f1b150-d1a9-4b90-8947-238bf1d87eb2.png)

It makes it easy to spot the weird empty loops we end up with too:
![image](https://user-images.githubusercontent.com/75337/153269736-75721198-70e9-456d-8cb0-d58d1c496ba5.png)

**Comment by benvanik** [comment]
(those shots are from dynamic_abs.mlir - which is just an elementwise abs on dense data - so the target would be a single loop and no pointer manipulation/striding/etc - don't expect us to get there in one step but good to know where we should be :)

**Comment by MaheshRavishankar** [comment]
There is a long thread that will eventually connect to this working

1) We move tile and distribute out of Flow and into codegen. That is taking view of dimensionality and making some early unrevertable decisions.
2) Then in Flow dispatches we can easily figure out cases where a generic op is essentially a 1D op. Then we collapse those dimensions with reshapes at the boundaries. This will remove all these loops.

I already have the code for (2) in another patch I am working on. (1) should follow next.

**Comment by MaheshRavishankar** [comment]
Adding a bit more detail here before assigning it to Okwan.

There are two kinds of dispatches in IREE, one where the op driving compilation in the dispatch is a GEMM, Convolution, etc. and one where the dispatch region contains just a generic op. Removing the loops generated by the compilation process in the former need to be done at the very end stage. There are several reasons for this, most important of which is that tensors + distributions dont gel that well, and the loop becomes semantically load bearing (see [this](https://discourse.llvm.org/t/representing-tiling-on-tensors-parallelism/4575) that has more info and a lot more other things). For such cases, removing the loops should be the last step in the code generation (and it might well turn out that we dont have all the information to remove/collapse such loops. We might have to live with those for now, we just dont have the constructs to linearize them after the fact currently).

For the second case , i.e. just a generic operation, flattening it makes sense in most common case.

```
#map0 = affine_map<(d0, d1) -> (d0, d1)>
%0 = linalg.generic {
    iterator_types = ["parallel", "parallel"]
    indexing_maps = [#map0, #map0, #map0]}
    ins(%in0, %in1 : tensor<?x?xf32>, tensor<?x?xf32>) outs(%out : tensor<?x?xf32>) { ... }
```

can be linearized.  to 

```
#map0 = affine_map<(d0) -> (d0)>
%0 = linalg.generic {
    iterator_types = ["parallel"]
    indexing_maps = [#map0, #map0, #map0]}
    ins(%in0, %in1 : tensor<?xf32>, tensor<?xf32>) outs(%out : tensor<?xf32>) { ... }
```

but some generics op are better off not being linearized this way.
For example
```
#map0 = affine_map<(d0, d1) -> (d0, d1)>
#map1 = affine_map<(d0, d1) -> (d0)>
%0 = linalg.generic {
    iterator_types = ["parallel", "parallel"]
    indexing_maps = [#map0, #map1, #map0]}
    ins(%in0, %in1 : tensor<?x?xf32>, tensor<?xf32>) outs(%out : tensor<?x?xf32>) { ... }
```

and

```
#map0 = affine_map<(d0, d1) -> (d0, d1)>
#map1 = affine_map<(d0, d1) -> (d1, d0)>
%0 = linalg.generic {
    iterator_types = ["parallel", "parallel"]
    indexing_maps = [#map0, #map1]}
    ins(%in0: tensor<?x?xf32>) outs(%out : tensor<?x?xf32>) { ... }
```

reason being that linearizing those result in the affine maps not being projected permutation which foils a lot of transformations downstream. 

Coming to the transformation for linearizing the iteration space itself, a lot of the code for this can probably reused from here https://github.com/llvm/llvm-project/blob/097035f308d1cdc00d71d3d216564fddffda0be7/mlir/lib/Dialect/Linalg/Transforms/ElementwiseOpFusion.cpp#L1466

**Comment by MaheshRavishankar** [comment]
Assigning to Okwan for now to start with the linearization of "some" generic ops to see if it helps the situation

**Comment by okkwon** [comment]
For the rescale tosa input, I don't see a bad graph. @benvanik would you mind attaching the dynamic_abs.mlir?

<img width="512" alt="image" src="https://user-images.githubusercontent.com/4314836/160461862-9b532ccb-5cf7-486e-a7a7-411ff793df2c.png">

**Comment by benvanik** [comment]
It's possible it's been improved for certain cases since this was filed - that'd be great! Did you try a rescale with several non-unit dimensions? You can track the dispatch region through and see what happens - if for example you don't get the same IR after dispatch region formation as in the issue comment then you won't see the same output, or you may find that there's now something inbetween that's flattening out the loops.

**Comment by okkwon** [comment]
>> several non-unit dimensions?
Would you give me an example?

BTW, `RemoveSingleIterationLoop` removes the outer loops related to the workgroup counters, which means there are outer loops in which the dispatch function is called, but in the dispatch function we are not supposed to see the outer loops (unless there is an transformation that sinks the outer loops into the dispatch function.)

**Comment by MaheshRavishankar** [comment]
Yeah, `RemoveSingleIterationLoop` was meant to help this a bit, but there is still the loops created while lowering to loops, or with second-level tiling. If it is not a trivially elidable loop, then these need linearization to be able to remove those loops (as well as associated index computation).

**Comment by benvanik** [comment]
yeah as mahesh mentions there's second level tiling that will still need loops inside of the workgroup - but for any example it's possible that the workgroup size is chosen such that the loops fold away - I suspect that's what changed between the original issue and now. The dynamic case mahesh posts won't be able to fold away. The key thing there is that for elementwise things we should be doing the second level tiling in the linearized spaces instead of the multi-dimensional space (for memory locality/loop overhead/etc reasons).

**Comment by okkwon** [comment]
Thanks Mahesh and Ben. I will take another look. 👍

**Comment by benvanik** [comment]
mspaint diagram time:
![image](https://user-images.githubusercontent.com/75337/160470478-b3d4b23a-dd04-4da0-9784-9186366af8b4.png)

With the intra-workgroup tiling based on the dimensions of the tensor we end up with the memory access pattern on the left, while a tiling based on a linearized space (even if partial) would give us the rhs. Much friendlier to the prefetcher and the entire cache hierarchy (less duplication in separate caches, shorter lifetimes in shared caches).

**Comment by MaheshRavishankar** [comment]
+1 to everything Ben mentions. To add some context. Some ops need the dimensionality information, and to enable fusion ops like generic ops are better to be reasoned about at higher dimensionality. But after that, its pure cost within a dispatch that has just a generic op. There are cases where we cannot do this (I have more details in a comment above). But wherever we can, we should linearize dimensions as much as possible. I suspect 90% of cases we hit are in the trivially linearizable form

**Comment by okkwon** [comment]
A few comments:
- We can do a tiling based on the row (aka. row partitioning) instead.
- Don't we need some kind of data affinity analysis?
- Controlling the tile size or how to partition may be more flexible than linearizing.

**Comment by benvanik** [comment]
For any operation in which we can linearize those will all have issues that linearizing doesn't and they're the kind of things we want to avoid :)

- Working with tensor<1024x2xf32> vs tensor<2x1024xf32> (or tensor<2x?x1024xf32> etc) should have zero performance difference - if you try to do anything based on the semantic dimensions of a tensor you will end up in a bad place. Worse, if any dims are dynamic you won't know. Extrapolate beyond simple ML ops and think of generic programs (what we produce after lots of frontend + linalg-level transforms); we don't want special cases for rank-1/rank-2/rank-3/rank-4 cases and failure on rank-5+ - all of this should be agnostic to what the user specified so long as the user-visible output data is the same.
- linalg.generic affine maps provide what we need (IIUC).
- Tile size changes alone don't help the problem here (lots of loop overhead) or what we want to help (memory subsystem-friendly access patterns) - both workgroup counts and tile sizes should be based on the data access patterns, not the other way around. Most of what we're talking about is improving the data access patterns so that the target-specific logic for work distribution have the best chance at picking something reasonable. Think of it like reducing the search space by dropping extra bits of entropy.

This issue shows an example of where trying to preserve high level information too low into the system causes issues. As a general rule we want to aggressively factor out user-level semantics that don't provide direct benefit to (or actively pessimize) subsequent layers in the system. The goal is to not need to ensure the entire system for all targets for all time has its heuristics consistent to avoid cases where we can't for example use SIMD instructions on large contiguous chunks of a tensor<1024x2xi8> but can on tensor<128x16xi8>, etc.

**Comment by benvanik** [comment]
(think of it like how memcpy doesn't care about the contents of the memory being copied - memcpy of int a[8][16] is the same as a[16][8] or a[128] - how memcpy chooses to execute is only based on how many bytes its copying/their alignment/etc, not the user-level semantics of those bytes)

**Comment by MaheshRavishankar** [comment]
> A few comments:
> 
> * We can do a tiling based on the row (aka. row partitioning) instead.

For a simple elementwise operation there is no reason to do that. It also has the effect of reducing the amount of parallelism exposed to threads. What if your iteration space is dynamically (or even statically for that matter) of size 1x1000 . If you partition along row, then the computation is serialized on a thread even though there is a lot of parallelism

> * Don't we need some kind of data affinity analysis?

That is what this comment is laying out (https://github.com/google/iree/issues/8053#issuecomment-1064767496)

> * Controlling the tile size or how to partition may be more flexible than linearizing.

At the flow level there is no tiling. Tiling happens in the backend, and these are separate axes. Linearization is always beneficial past dispatch region formation. It exposes the maximum amount of parallelism. Then you can control the tile sizes for the remaining dimensions.

> * Linearization can be done after the tiling decision; this would separate out the two things.

Again, different axes. Once you tile, its very hard to linearize (if you are talking about the loops). That's another thing we could do, (that something being looked into) but again is orthogonal. At this level, if you can linearize, anything you do after is only going to benefit from it (provided you linearize based on the comment https://github.com/google/iree/issues/8053#issuecomment-1064767496)

**Comment by okkwon** [comment]
> Working with tensor<1024x2xf32> vs tensor<2x1024xf32> (or tensor<2x?x1024xf32> etc) should have zero performance difference

I understand the expectation, but it depends on how the work is distributed to the computational units. For example, due to a data dependency, the previous work might have been done with a column partitioning. (This is gonna be more relevant with affine-based HPC applications rather than ML application.) The proposed linearization is equivalent to a row partitioning. It implicitly dictates how the tiling is done since it happens before tiling. In this case, the communication (or data movement pattern) between the column partitioning and the row partitioning causes a lot of data movement. This is the All-to-All communication pattern. (I am extending my view of the problem to the distributed memory systems, which we may want to support later with multi-gpu or multi-node.) If we did a column partitioning for the second operation, there would be no communication at all.

> linalg.generic affine maps provide what we need (IIUC)

After the linearization transform, we would have different affine map. In particular, having a different iteration space representation might make the dispatch fusion more difficult.

> Tile size changes alone don't help the problem here ...

I agree. I am trying to separate out the loop overhead reduction from tiling. Again, linearizing the element-wise generic op affects how the tiling happens, and it makes me pause and think..

**Comment by MaheshRavishankar** [comment]
> 
> After the linearization transform, we would have different affine map. In particular, having a different iteration space representation might make the dispatch fusion more difficult.

That really depends. It could also open up more avenues for fusion. If you had something with an iteration space of 20x30 and another with an iteration space of 30x20, earlier they are not easy to fuse, after linearization, they are now potentially fusable.

**Comment by okkwon** [comment]
Keeping the discussion noted, I will go ahead and implement the linearization. We can revisit and enhance once we have more concrete problems. Thanks Ben and Mahesh!

**Comment by MaheshRavishankar** [comment]
> Keeping the discussion noted, I will go ahead and implement the linearization. We can revisit and enhance once we have more concrete problems. Thanks Ben and Mahesh!

Your point about distributed memory system is valid. When we cross that bridge we can decide how and where the transformation here would trigger. It will probably still have a place "post distribution".

**Comment by grypp** [comment]
Hi all, #11295  could linearize `linalg.generic` to 1D on this issue. Please let me know if you have other uses case so I could try.
